src/hardware/disk_noise.cpp

```diff
@@ -0,0 +1,479 @@
+			                    device->spin_sample_.size() ||
+			            device->loop_spin_sample_)) {
+				mixed_l += scale_sample(
+				        device->spin_sample_[device->spin_pos_],
+				        device->spin_volume_,
+				        mix_scale);
+				mixed_r += scale_sample(
+				        device->spin_sample_[device->spin_pos_++],
+				        device->spin_volume_,
+				        mix_scale);
+
+				// Loop the spin sound if enabled. Used for
+				// persistent HDD noise Not used for floppy
+				// noise because motor should stop after r/w
+				// operations aredone
+				if (device->spin_pos_ >= device->spin_sample_.size() &&
+				    device->loop_spin_sample_) {
+					device->spin_pos_ = 0;
+				}
+			}
+
+			if (!device->current_seek_sample_.empty() &&
+			    device->seek_pos_ + 1 <
+			            device->current_seek_sample_.size()) {
+				mixed_l += scale_sample(
+				        device->current_seek_sample_[device->seek_pos_],
+				        device->seek_volume_,
+				        mix_scale);
+				mixed_r += scale_sample(
+				        device->current_seek_sample_[device->seek_pos_++],
+				        device->seek_volume_,
+				        mix_scale);
+			}
+
+			out[i * 2] += mixed_l;
+			out[i * 2 + 1] += mixed_r;
+		}
+
+		if (!device->current_seek_sample_.empty() &&
+		    device->seek_pos_ >= device->current_seek_sample_.size()) {
+			device->current_seek_sample_.clear();
+		}
+	}
+
+	mix_channel_->AddSamples_sfloat(frames, out.data());
+}
+
+void DiskNoiseDevice::Shutdown()
+{
+	std::lock_guard<std::mutex> lock(device_mutex_);
+	active_devices_.erase(std::remove(active_devices_.begin(),
+	                                  active_devices_.end(),
+	                                  this),
+	                      active_devices_.end());
+
+	if (active_devices_.empty() && mix_channel_) {
+		mix_channel_->Enable(false);
+		MIXER_DeregisterChannel(mix_channel_);
+		mix_channel_.reset();
+	}
+}
+
+std::unique_ptr<DiskNoiseDevice> floppy_noise = nullptr;
+std::unique_ptr<DiskNoiseDevice> hdd_noise    = nullptr;
+
+void DISKNOISE_Init(Section* section)
+{
+	assert(section);
+	const auto prop = static_cast<Section_prop*>(section);
+
+	const bool enable_floppy_disk_noise = prop->Get_bool("floppy_disk_noise");
+	const bool enable_hard_disk_noise = prop->Get_bool("hard_disk_noise");
+
+	const auto spin_up = "hdd_spinup.flac";
+	const auto spin    = "hdd_spin.flac";
+	std::vector<std::string> hdd_seek_samples;
+	for (int i = 1; i <= 9; ++i) {
+		hdd_seek_samples.push_back("hdd_seek" + std::to_string(i) + ".flac");
+	}
+
+	const auto floppy_spin_up = "fdd_spinup.flac";
+	const auto floppy_spin    = "fdd_spin.flac";
+	std::vector<std::string> floppy_seek_samples;
+	for (int i = 1; i <= 9; ++i) {
+		floppy_seek_samples.push_back("fdd_seek" + std::to_string(i) + ".flac");
+	}
+
+	constexpr float hdd_spin_volume = 0.4f;
+	constexpr float hdd_seek_volume = 0.2f;
+
+	constexpr float floppy_spin_volume = 0.2f;
+	constexpr float floppy_seek_volume = 0.6f;
```

Comment:
Scaling the output to a lower volume is a good idea because we're using 16-bit samples. This way we can maintain healthy signal levels at the sample level 👍🏻 

But I strongly suggest to use a _single_ scaling factor for all samples. Then whoever authors the samples must ensure they have the appropriate perceptual loudness relative to each other as a set. Otherwise authoring new sample sets will be very difficult due to these hardcoded per-sample scaling factors (the author would need to pre-apply these volume scalings at the authoring stage so the "come out just right" after these scalings).